src/Moq/Mock.Generic.cs

```diff
@@ -147,6 +147,21 @@ public Mock(MockBehavior behavior, params object[] args)
 			this.CheckParameters();
 		}
 
+		/// <summary>
+		///   Initializes an instance of the mock using the given constructor call including its
+		///   argument values and with a specific <see cref="MockBehavior"/> behavior.
+		/// </summary>
+		/// <param name="constructorExpression">Lambda expression that creates an instance of <typeparamref name="T"/> T.</param>
+		/// <param name="behavior">Behavior of the mock.</param>
+		/// <example>
+		/// <code>var mock = new Mock&lt;MyProvider&gt;(() => new MyProvider(someArgument, 25));</code>
+		/// </example>
+		[SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors")]
+		public Mock(Expression<Func<T>> constructorExpression, MockBehavior behavior = MockBehavior.Strict)
```

Comment:
Please put the `MockBehavior` parameter in first position so that the signature is closer to the existing `public Mock(MockBehavior behavior, params object[] args)`. (Since it's an optional parameter, that means you'll need to add another constructor overload.)

Reply:
Did the change locally, there is now an ambiguity when using `new Mock<Foo>(null)` between `Mock<T>.Mock(params object[])` and `Mock<T>.Mock(Expression<Func<T>>)`.

I have no strong preference between what I did and what you suggested... but given that it introduces an ambiguity I see no other alternative than keeping what I did.

What do you think?

Reply:
Pushed the change in a separate commit so you can see the ambiguity issue...

Reply:
@stakx I'm afraid you were waiting until the PR status gets green again. It's red to show the issue when reverting the order of parameters as suggested. If you have some time please have a look... and sorry if you were off!

Reply:
@frblondin, sorry for not having replied yet, I thought that's why the build status was currently marked as failed.

> Did the change locally, there is now an ambiguity when using `new Mock<Foo>(null)` between `Mock<T>.Mock(params object[])` and `Mock<T>.Mock(Expression<Func<T>>)`.

If the C# compiler didn't refuse to compile ambiguous calls i.e. if it just chose one of the two possible methods, we could have made sure ourselves that a call with `null` gets redirected to the `params object[]` overload (preexisting behavior).

But since it completely refuses to compile ambiguous calls, I see no other alternative to what you're suggesting: revert the parameter order back to how you initially had it.

My apologies for not realizing earlier that we'd run into this!

Reply:
> I see no other alternative to what you're suggesting: revert the parameter order back to how you initially had it.

Reverted the last commit.

> My apologies for not realizing earlier that we'd run into this!

No worries, it was really tricky to forecast this type of compiler ambiguity 🔎 .